details/impersonation.md

```diff
@@ -0,0 +1,144 @@
+      users: ["*"]              // Who can be impersonated? (User names or patterns)
+      auth_key: admin1:pass     // Authentication rule required to impersonate (any authentication rule can be used here)
+
+    - impersonator: admin2
+      users: ["dev2"]
+      ldap_authentication: "ldap1"
+```
+
+In the example above, we see that we have two impersonators: `admin1` and `admin2`. The first one can impersonate any user (`*`) and they are able to authenticate using basic auth (`admin1:pass`). The second impersonator can impersonate only `dev2` user. They will be authenticated using `ldap1` connector.
+
+When an impersonator passes wrong credentials ROR will tell Kibana that impersonation is not allowed.
+
+#### Defining mocks of the external services (optional)
+
+ROR has many sophisticated authentication & authorization methods. Some of them are based on external systems like LDAP. A problem with such systems according to the impersonation feature is that, the systems don't support it by default, or the configuration is difficult, or they don't support it at all. 
+
+That's why we decided to solve it totally differently - using mocks. [Wikipedia](https://en.wiktionary.org/wiki/mock) defines `mock` as `an imitation, usually of lesser quality.` And in case of external authentication system we are going provide an imitation of it that tell ACL what users can be authenticated by it. When we consider an authorization service, a mock of it will return the ACL users with their roles in the service. And this is enough for ROR to support impersonation. 
+
+How does ROR use the mocks? Let's suppose we have `ldap_auth` rule. When ROR processes the rule, it:
+* asks a given LDAP if the username can be authenticated with a given password, and if they can ...
+* asks LDAP to list what groups the user belongs to
+
+In the impersonation case, it looks pretty much the same. Just we won't call any LDAP server, rather, the mock will provide the required information (no passwords required):
+* asks the mock if the username exists, and if it does ...
+* asks the mock to tell what groups the user belongs to
+
+**⚠️ IMPORTANT:** When some external service is not mocked, ROR might return to Kibana information that the impersonation is not supported. It's better to always define all mocks, to avoid the "not supported impersonation" ES response.
+
+How to define mocks in Kibana is described [here](../examples/impersonation/external-services-mocks-ui.md).
+
+#### Impersonating a chosen user
+
+Now, when we have configured Test Settings and External Services Mocks, we can try to impersonate a user. In ROR Settings, user can be:
+* provided statically (defined in the settings),
+* provided dynamically:
+   * from external, dependant systems (like LDAP) - we mock them
+   * from upstream systems (eg. through headers) - they are unknown
+
+It means that we pick the users defined in Settings or Mocks, but also we can enter the username and try to impersonate such user.
```

Comment:
This is followup to the previous comment and maybe it stems from my limited knowledge of how ROR_ES works, but I don't understand this. Who are these users not defined in settings nor in mocks? 🤔

Reply:
In ROR settings, in general, a user can be:
* defined statically (in some cases ROR doesn't know the usernames because eg. they are hashed)
* be defined in an external service (eg. LDAP)
* be sent to ROR (with JWT or header)

IMO there is no sense to write here about it because it's rather related to the ROR rules.